common/persistence/sql/sqlClusterMetadataManager.go

```diff
@@ -41,56 +43,49 @@ type sqlClusterMetadataManager struct {
 
 var _ p.ClusterMetadataStore = (*sqlClusterMetadataManager)(nil)
 
-func (s *sqlClusterMetadataManager) InitializeImmutableClusterMetadata(request *p.InternalInitializeImmutableClusterMetadataRequest) (*p.InternalInitializeImmutableClusterMetadataResponse, error) {
-	resp, err := s.GetImmutableClusterMetadata()
+func (s *sqlClusterMetadataManager) GetClusterMetadata() (*p.InternalGetClusterMetadataResponse, error) {
+	row, err := s.db.GetClusterMetadata()
 
 	if err != nil {
-		if _, ok := err.(*serviceerror.NotFound); ok {
-			// If we have received EntityNotExistsError, we have not yet initialized
-			return s.InsertImmutableDataIfNotExists(request)
-		}
-		return nil, err
+		return nil, convertCommonErrors("GetClusterMetadata", err)
 	}
 
-	// Return our get result if we didn't need to initialize
-	return &p.InternalInitializeImmutableClusterMetadataResponse{
-		PersistedImmutableMetadata: resp.ImmutableClusterMetadata,
-		RequestApplied:             false,
+	return &p.InternalGetClusterMetadataResponse{
+		ClusterMetadata: p.NewDataBlob(row.Data, row.DataEncoding),
+		Version:         row.Version,
 	}, nil
-
 }
 
-func (s *sqlClusterMetadataManager) InsertImmutableDataIfNotExists(request *p.InternalInitializeImmutableClusterMetadataRequest) (*p.InternalInitializeImmutableClusterMetadataResponse, error) {
-	// InsertIfNotExists is idempotent and silently fails if already exists.
-	// Assuming that if we make it here, no out-of-band method or tool is deleting the db row
-	//	in between the Get above and Insert below as that would violate the immutability guarantees.
-	// Alternative would be to make the insert non-idempotent and detect insert conflicts
-	// or even move to a lock mechanism, but that doesn't appear worth the extra lines of code.
-	_, err := s.db.InsertIfNotExistsIntoClusterMetadata(&sqlplugin.ClusterMetadataRow{
-		ImmutableData:         request.ImmutableClusterMetadata.Data,
-		ImmutableDataEncoding: request.ImmutableClusterMetadata.Encoding.String(),
+func (s *sqlClusterMetadataManager) SaveClusterMetadata(request *p.InternalSaveClusterMetadataRequest) (bool, error) {
+	err := s.txExecute("SaveClusterMetadata", func(tx sqlplugin.Tx) error {
+		oldClusterMetadata, err := tx.GetClusterMetadata()
```

Comment:
there are 2 ways to conditionally update the record
1. update something if some conditional is met, like the implementation if Cassandra
2. lock the record, like https://github.com/temporalio/temporal/blob/5cc53ee68036bb32d15241ed011b36f81950cf92/common/persistence/sql/sqlplugin/mysql/execution.go#L51-L55

according to the current implementation, adding a lock function seems to be the solution

Reply:
Addressed.